dbms/src/Common/ConcurrentIOQueue.h

```diff
@@ -26,132 +26,127 @@ namespace DB
     {
-        return mpmc_queue.tryPush(std::forward<U>(data));
+        std::lock_guard lock(mu);
+
+        if unlikely (status == MPMCQueueStatus::CANCELLED)
+            return MPMCQueueResult::CANCELLED;
+        if unlikely (status == MPMCQueueStatus::FINISHED)
+            return MPMCQueueResult::FINISHED;
+
+        if (queue.size() >= capacity)
+            return MPMCQueueResult::FULL;
+
+        pushFront(std::forward<U>(data));
+        return MPMCQueueResult::OK;
     }
 
     /// Non-blocking function.
     /// Besides all conditions mentioned at `push`, `nonBlockingPush` will still return OK if queue is `NORMAL` and full.
-    /// The obj that exceeds its capacity will be stored in remaining_objs and wait for the next pop/tryPop to trigger pushRemains.
+    /// The obj that exceeds its capacity will still be stored in queue.
     MPMCQueueResult nonBlockingPush(T && data)
     {
-        auto res = mpmc_queue.tryPush(std::move(data));
-        if (res == MPMCQueueResult::FULL)
-        {
-            // Double check if this queue is full.
-            std::lock_guard lock(mu);
-            res = mpmc_queue.tryPush(std::move(data));
-            if (res == MPMCQueueResult::FULL)
-            {
-                // obj that exceeds its capacity will be stored in remaining_objs.
-                // And then waiting for the next pop/tryPop to trigger pushRemains.
-                remaining_objs.push_front(std::move(data));
-                return MPMCQueueResult::OK;
-            }
-        }
-        return res;
+        std::lock_guard lock(mu);
+
+        if unlikely (status == MPMCQueueStatus::CANCELLED)
+            return MPMCQueueResult::CANCELLED;
+        if unlikely (status == MPMCQueueStatus::FINISHED)
+            return MPMCQueueResult::FINISHED;
+
+        pushFront(std::move(data));
+        return MPMCQueueResult::OK;
     }
 
     MPMCQueueResult pop(T & data)
     {
-        auto res = mpmc_queue.pop(data);
-        switch (res)
+        std::unique_lock lock(mu);
+        reader_head.wait(lock, [&] { return !queue.empty() || (unlikely(status != MPMCQueueStatus::NORMAL)); });
+
+        if ((likely(status != MPMCQueueStatus::CANCELLED)) && !queue.empty())
         {
-        case MPMCQueueResult::FINISHED:
-            // when finished, we still should pop the remaining_objs.
-            return popRemains(data) ? MPMCQueueResult::OK : MPMCQueueResult::FINISHED;
-        case MPMCQueueResult::OK:
-            pushRemains();
-        default:
-            return res;
+            data = popBack();
+            return MPMCQueueResult::OK;
+        }
+
+        switch (status)
+        {
+        case MPMCQueueStatus::NORMAL:
+            return MPMCQueueResult::EMPTY;
+        case MPMCQueueStatus::CANCELLED:
+            return MPMCQueueResult::CANCELLED;
+        case MPMCQueueStatus::FINISHED:
+            return MPMCQueueResult::FINISHED;
         }
     }
 
     MPMCQueueResult tryPop(T & data)
     {
```

Comment:
It seems the major performance regression is caused by double-checking. If the mutex is acquired first without double-checking, I think the performance difference is not so much.

Reply:
I think so too

Reply:
But there should be a double gap, just like `isFull`